### DISCUSSION TITLE
What's going on with `all_node_cuts()` and `Set().update()`?

### DISCUSSION THREAD

**Discussion by Cerebus** [networkx/networkx#7533] [Q&A]
Calling `next(nx.all_node_cuts(graph))` can take a long time.  How long it takes seems to be a function of calling `Set().update()`.  For example, on `g = nx.grid_2d_graph(8, 8)`:

```
Mon Jul  1 11:42:37 2024    /tmp/cuts8

         507929371 function calls (507907839 primitive calls) in 338.136 seconds

   Ordered by: cumulative time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.000    0.000  338.136  338.136 {built-in method builtins.exec}
        1    0.000    0.000  338.136  338.136 <string>:1(<module>)
  21522/1    0.012    0.000  338.136  338.136 {built-in method builtins.next}
        1  113.506  113.506  338.136  338.136 kcutsets.py:23(all_node_cuts)
215741086  136.516    0.000  165.318    0.000 {method 'update' of 'set' objects}
...
```

This is consistent in all my profiling; `Set().update()` consumes about half the cumulative time of `all_node_cuts()`.

However, the overall *pattern* of *how often* `Set().update()` is called is ... weird ... and doesn't appear consistent with Kavensky's algorithm.  For 2D grids of size (n, n), n=[1..13]:

Python 3.8.10 and 3.12.4:
```
n     ncalls  tottime  percall  cumtime  percall filename:lineno(function)
1          2    0.000    0.000    0.000    0.000 {method 'update' of 'set'}
2         26    0.000    0.000    0.000    0.000 {method 'update' of 'set'}
3        141    0.000    0.000    0.000    0.000 {method 'update' of 'set'}
4        121    0.000    0.000    0.000    0.000 {method 'update' of 'set'}
5       1113    0.001    0.000    0.001    0.000 {method 'update' of 'set'}
6       6429    0.003    0.000    0.004    0.000 {method 'update' of 'set'}
7        352    0.000    0.000    0.001    0.000 {method 'update' of 'set'}
8  215741086  103.683    0.000  130.149    0.000 {method 'update' of 'set'}
9     865597    0.382    0.000    0.518    0.000 {method 'update' of 'set'}
10       709    0.001    0.000    0.002    0.000 {method 'update' of 'set'}
11       856    0.001    0.000    0.002    0.000 {method 'update' of 'set'}
12      1017    0.002    0.000    0.003    0.000 {method 'update' of 'set'}
13      1192    0.002    0.000    0.004    0.000 {method 'update' of 'set'}
```

Meanwhile the same method counts are *substantially* different under Python 3.7:

Python 3.7.17:
```
n    ncalls  tottime  percall  cumtime  percall filename:lineno(function)
1         2    0.000    0.000    0.000    0.000 {method 'update' of 'set'}
2        26    0.000    0.000    0.000    0.000 {method 'update' of 'set'}
3       141    0.000    0.000    0.000    0.000 {method 'update' of 'set'}
4       159    0.000    0.000    0.000    0.000 {method 'update' of 'set'}
5       240    0.000    0.000    0.000    0.000 {method 'update' of 'set'}
6    117587    0.036    0.000    0.048    0.000 {method 'update' of 'set'}
7   2523211    0.861    0.000    1.142    0.000 {method 'update' of 'set'}
8  52035762   20.267    0.000   26.526    0.000 {method 'update' of 'set'}
9   3566529    1.247    0.000    1.728    0.000 {method 'update' of 'set'}
10 17194765    6.799    0.000    9.323    0.000 {method 'update' of 'set'}
11  4885322    2.136    0.000    2.976    0.000 {method 'update' of 'set'}
```

FWIW, stats from cProfile on NetworkX 2.6.3 b/c that's the latest version for Py3.7.  Current NetworkX release is identical (verified empirically, but AFAICT nothing substantive has changed in `kcutsets.py` in the interim.  I skimmed the Py3.8 changelog and nothing jumped out at me, but I need to look at it in more detail.

Am I doing something wrong here, or is there seriously something janky happening?

Driver code:
```python
import networkx as nx
import cProfile, pstats

for n in range(1, 14):
    g = nx.grid_2d_graph(n, n)
    c = nx.all_node_cuts(g)
    cProfile.runctx("next(c)", {"c": c}, {}, filename=f"/tmp/cuts{n}")
    p = pstats.Stats(f"/tmp/cuts{n}")
    p.strip_dirs().print_stats("{method 'update' of 'set' objects}")
```

**Comment by dschult** [comment]
It looks like what you are measuring is how much work is being done as a function of size. One difficult aspect of network algorithms is that small changes in the graph can make huge changes in the run-time. And it looks like 8x8 grid graphs are just a lot harder to find cuts for than 7x7 or 9x9. I'm not surprised by the n=8 case. 8 is a very "symmetry" inducing number and an 8 by 8 grid would have lots of symmetries too. So it makes sense to me that it would take a lot longer to do. But for this algorithm I don't know how network structure affects run-time, so it may have nothing to do with symmetries.

It might help you understand more about the algorithm if you figure out how many possible "potential cuts" there are for each of these cases -- using some counting technique. It is likely to be searching through a large number of potential cuts to find one that is actually a cut. :)
